nemo/collections/nlp/modules/common/transformer/reduction_encoders.py

```diff
@@ -0,0 +1,174 @@
+
+        if self.hidden_init_method not in self.supported_init_methods:
+            raise ValueError(
+                "Unknown hidden_init_method = {hidden_init_method}, supported methods are {supported_init_methods}".format(
+                    hidden_init_method=self.hidden_init_method, supported_init_methods=self.supported_init_methods,
+                )
+            )
+
+        # self-attention encoder
+        layer = TransformerEncoder(
+            num_layers=num_layers,
+            hidden_size=hidden_size,
+            inner_size=inner_size,
+            mask_future=mask_future,
+            num_attention_heads=num_attention_heads,
+            attn_score_dropout=attn_score_dropout,
+            attn_layer_dropout=attn_layer_dropout,
+            ffn_dropout=ffn_dropout,
+            hidden_act=hidden_act,
+            pre_ln=pre_ln,
+            pre_ln_final_layer_norm=pre_ln_final_layer_norm,
+        )
+        self.self_att_layers = torch.nn.ModuleList([copy.deepcopy(layer) for _ in range(hidden_blocks)])
+
+        self.pooling = self._build_pooling_module()
+
+    def _build_pooling_module(self):
+        """
+        Returns pooling module.
+        Allows to override for child classes.
+        """
+        return torch.nn.MaxPool1d(kernel_size=2, stride=2)
+
+    @property
+    def supported_init_methods(self):
+        return ["default"]
+
+    @property
+    def hidden_steps(self):
+        return self._hidden_steps
+
+    @property
+    def hidden_blocks(self):
+        return self._hidden_blocks
+
+    @property
+    def hidden_init_method(self):
+        return self._hidden_init_method
+
+    def forward(self, encoder_states, encoder_mask):
+        """
+        Args:
+            encoder_states: output of the encoder (B x L_enc x H)
+            encoder_mask: encoder inputs mask (B x L_enc)
+        """
+        # initialize hidden state
+        hidden_mask = encoder_mask
+        hidden_states = encoder_states
+
+        # apply block (self-attention, max-pool) multiple times
+        for self_att in self.self_att_layers:
+            residual = hidden_states
+
+            # self-attention over hidden
+            hidden_states = self_att(encoder_states=hidden_states, encoder_mask=hidden_mask)
+
+            hidden_states += residual
+
+            # max pool reduction if possible
+            if hidden_states.shape[1] >= self.hidden_steps:
+                # max pool hidden states
+                hidden_states = hidden_states.permute(0, 2, 1)
+                hidden_states = self.pooling(hidden_states)
+                hidden_states = hidden_states.permute(0, 2, 1)
+
+                # max pool mask
+                hidden_mask = (
+                    self.pooling(hidden_mask.unsqueeze(0).type_as(hidden_states)).squeeze(0).type_as(hidden_mask)
+                )
+
+        return hidden_states, hidden_mask
+
+
+class AveragePoolingEncoder(torch.nn.Module):
```

Comment:
AvgPool does not seem to have methods (like forward) defined. Did you intend to inherit from `MaxPoolingEncoder`?
Also, why not just have 1 class, say PoolingEncoder, that takes as param the type of pooling and chooses the nn class to call (Avg or max) accordingly?

Reply:
Great feedback. Done.